File: evd_ros_backend/evd_ros_core/src/evd_script/program/task.py
from .primitive import Primitive


class Task(Primitive):

    '''
    Data structure methods
    '''

    def __init__(self, primitives=[], type='', name='', uuid=None, parent=None, append_type=True):

        self._primitives = []

        super(Task,self).__init__(
            type='task.'+type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type)

        self.primitives = primitives

    def to_dct(self):
        msg = super(Task,self).to_dct()
        msg.update({
            'primitives': [p.to_dct() for p in self.primitives]
        })
        return msg

    @classmethod
    def from_dct(cls, dct):
        from ..utility_functions import NodeParser

        return cls(
            name=dct['name'],
            type=dct['type'],
            append_type=False,
            uuid=dct['uuid'],
            primitives=[NodeParser(p) for p in dct['primitives']])

    '''
    Data accessor/modifier methods
    '''

    @property
    def parent(self):
        return self._parent

    @parent.setter
    def parent(self, value):
        if self._parent != value:

            self.remove_from_cache()
            self._parent = value
            self.add_to_cache()

            self.updated_attribute("parent","set")

    @property
    def primitives(self):
        return self._primitives

    @primitives.setter
    def primitives(self, value):
        if self._primitives != value:
            for p in self._primitives:
                p.remove_from_cache()

            self._primitives = value
            for p in self._primitives:
                p.parent = self

            self.updated_attribute('primitives','set')

    def add_primitive(self, prm):
        prm.parent = self
        self._primitives.append(prm)
        self.updated_attribute('primitives','set')

    def reorder_primitives(self, uuid, shift):
        idx = None
        for i in range(0,len(self._primitives)):
            if self._primitives[i].uuid == uuid:
                idx = i
                break

        if idx != None:
            shiftedIdx = idx + shift
            if shiftedIdx < 0 or shiftedIdx >= len(self._primitives):
                raise Exception("Index out of bounds")

            copy = self._primitives.pop(idx)
            self._primitives.insert(shiftedIdx,copy) #TODO check to make sure not off by one

            self.updated_attribute('primitives','reorder')

    def delete_primitive(self, uuid):
        delIdx = None
        for i in range(0,len(self._primitives)):
            if self._primitives[i].uuid == uuid:
                delIdx = i
                break

        if delIdx != None:
            self._primitives.pop(i).remove_from_cache()
            self.updated_attribute('primitives','delete')

    def get_primitive(self, uuid):
        for p in self.primitives:
            if p.uuid == uuid:
                return p
        return None

    def set(self, dct):
        from ..utility_functions import NodeParser

        if 'primitives' in dct.keys():
            self.primitives = [NodeParser(p) for p in dct['primitives']]

        super(Task,self).set(dct)

    '''
    Cache methods
    '''

    def remove_from_cache(self):
        for p in self._primitives:
            p.remove_from_cache()

        super(Task,self).remove_from_cache()

    def add_to_cache(self):
        for p in self._primitives:
            p.add_to_cache()

        super(Task,self).add_to_cache()

    '''
    Children methods
    '''

    def delete_child(self, uuid):
        success = True

        if uuid in [p.uuid for p in self.primitives]:
            self.delete_primitive(uuid)
        else:
            success = False

        return success

    '''
    Update Methods
    '''

    def late_construct_update(self):

        for p in self.primitives:
            p.late_construct_update()

        super(Task,self).late_construct_update()

    def deep_update(self):

        for p in self.primitives:
            p.deep_update()

        super(Task,self).deep_update()

        self.updated_attribute('primitives','update')

    def shallow_update(self):
        super(Task,self).shallow_update()

        self.updated_attribute('primitives','update')

File: evd_authoring_app/src/model/evdScript/program/task.js
import { Primitive } from './primitive';
import { NodeParser } from '../utilityFunctions';


export class Task extends Primitive {

    /*
    * data structure methods
    */

    static typeString() {
        return 'task.';
    }

    static fullTypeString() {
        return Primitive.fullTypeString() + Task.typeString();
    }

    constructor(primitives=[], type='', name='', uuid=null, parent=null, appendType=true) {
        this._primitives = [];

        super(
            type= (appendType) ? 'task.'+type : type,
            name= name,
            uuid= uuid,
            parent= parent,
            appendType= appendType
        );

        this.primitives = primitives;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            primitives: this.primitives.map(p => p.toDict())
        };
        return msg;
    }

    static fromDict(dct) {
        return new Task(
            primitives= dct.primitives.map(p => NodeParser(p)),
            type= dct.type,
            name= dct.name,
            uuid= dct.uuid,
            appendType= false
        );
    }

    /*
    * data accessor/modifier methods
    */

    get parent() {
        return this._parent;
    }

    set parent(value) {
        if (this._parent !== value) {
            this.removeFromCache();
            this._parent = value;
            this.addToCache();

            this.updatedAttribute('parent','set');
        }
    }

    get primitives() {
        return this._primitives;
    }

    set primitives(value) {
        if (this._primitives !== value) {
            for (const p in this._primitives) {
                p.removeFromCache();
            }

            this._primitives = value;
            for (const p in this._primitives) {
                p.parent = this;
            }
        }
    }

    addPrimitive(primitive) {
        primitive.parent = this;
        this._primitives.push(primitive);
        this.updatedAttribute('primitives','add')
    }

    insertPrimitive(primitive, idx) {
        primitive.parent = this;
        this._primitives.splice(idx,0,primitive);
        this.updatedAttribute('primitives','add')
    }

    reorderPrimitives(uuid, shift) {
        const idx = this.findPrimitiveIndex(uuid);

        if (idx !== null) {
            let shiftedIdx = idx + shift;
            if (shiftedIdx < 0 || shiftedIdx >= this._primitives.length) {
                throw new Error('Index out of bounds');
            }

            const copy = this._primitives.splice(idx, 1);
            this._primitives.splice(shiftedIdx, 0, copy);
            this.updatedAttribute('primitives','reorder');
        }
    }

    deletePrimitive(uuid) {
        const idx = this.findPrimitiveIndex(uuid);

        if (idx !== null) {
            this._primitives.splice(idx, 1);
            this.updatedAttribute('primitives','delete',uuid);
        }
    }

    getPrimitive(uuid) {
        const idx = this.findPrimitiveIndex(uuid);
        if (idx !== null) {
            return this._primitives[idx];
        } else {
            return null;
        }
    }

    findPrimitiveIndex(uuid) {
        let idx = null;
        for (let i=0; i<this._primitives.length; i++) {
            if (this._primitives[i].uuid === uuid) {
                idx = i;
                break;
            }
        }
        return idx;
    }

    set(dct) {

        if ('primitives' in dct) {
            this.primitives = dct.primitives.map(p => NodeParser(p));
        }

        super.set(dct);
    }

    /*
    * cache methods
    */

    removeFromCache() {
        this.primitives.forEach(p => p.removeFromCache());
        super.removeFromCache();
    }

    addToCache() {
        this.primitives.forEach(p => p.addToCache());
        super.addToCache();
    }

    /*
    * children methods
    */

    deleteChild(uuid) {
        let success = true;

        if (this.findPrimitiveIndex(uuid) !== null) {
            this.deletePrimitive(uuid);
        } else {
            success = false;
        }

        return success;
    }

    /*
    * Update methods
    */

    lateConstructUpdate() {
        this.primitives.forEach(p => p.lateConstructUpdate());
        super.lateConstructUpdate();
    }

    deepUpdate() {
        this.primitives.forEach(p => p.deepUpdate());
        super.deepUpdate();
        
        this.updatedAttribute('primitives','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('primitives','update');
    }
}